### PR TITLE
File upload widget

### PR DESCRIPTION
**Pull request by robinr**
File Upload Widget feature as follows:-
a) Settings Vuex store which allows have values of file_count(number of files uploaded)
    max_file_count (max number of files uploaded)
b) b-progress bar (progressive bar controlled by the increments of file_count)
c) A Green Check mark at the end when file_count equals max_file_count

### CODE REVIEW

**Comment by robinr** [comment]
Note: "basic-upload-lastpoint-base" screenshoot with E2E testing completes the code coverage of the function which renders the 'Check' mark which is cleared via function call back.

File: components/status/UploadFilesWidget.vue [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+  height: 4px;
+  top: 20px;
+  left: 281px;
+  visibility: visible;
+  background-color: #4c4c4c;
+  z-index: 7;
+}
+
+#upload-progress-bar > .bg-success {
+  background-color: #00c46f !important;
```
Comment (eli88fine) [inline components/status/UploadFilesWidget.vue:172]: is the `!important` necessary here? is there another way to achieve this without using `!important`?  The Mozilla documentation and other places caution that `!important` should be very rarely used and you should try your best to use the CSS inheritence rules instead

File: tests/unit/components/status/UploadFilesWidgets.spec.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+      store,
+      localVue,
+    });
+
+    const target_upload_file_count_container = wrapper.find(
+      ".span__upload-file-count-container"
+    );
+    expect(target_upload_file_count_container.text()).toStrictEqual("500/900");
+  });
+  test("Given that the uploaded_files is equal to max_file_count, When complete upload is successful, Then the 'check mark' is visible indicating the completion of file upload activity", async () => {
```
Comment (eli88fine) [inline tests/unit/components/status/UploadFilesWidgets.spec.js:57]: For this test, commiting uploaded_files=900 is the trigger that causes the action being tested to occur isn't it?  Should that be in the `When` section, or am I misunderstanding your code and/or test?

File: tests/unit/components/status/UploadFilesWidgets.spec.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    );
+    expect(target_upload_custom_check_mark.isVisible()).toBe(true);
+    const target_upload_custom_check_mark_color = wrapper.find(
+      ".svg-inline-fa-check-path"
+    );
+    expect(
+      target_upload_custom_check_mark_color.attributes("fill")
+    ).toStrictEqual("#00c46f");
+  });
+  test("Given that the uploaded_files is less than max_file_count, When the upload of files is about 88%, Then the 'progress bar' is visible indicating and its less than the max 110px max length of progress bar attribute", async () => {
```
Comment (eli88fine) [inline tests/unit/components/status/UploadFilesWidgets.spec.js:82]: In general, if Vuex commits happen prior to mounting the component, then they are part of the Arrange/Given section of a test.  Here it looks like you are testing the behavior of the component When it mounts

File: tests/unit/store/settings.spec.js [hunk 4]
```diff
@@ -25,6 +25,12 @@ describe("store/settings", () => {
       .customer_account_ids;
     expect(array_of_customerids.length).toEqual(0);
   });
+  test("When imported from the dist file, Then the file_count and max_file_count is zero 0 as with no value assigned", () => {
```
Comment (eli88fine) [inline tests/unit/store/settings.spec.js:28]: This test is not importing from the dist file.  The `store` variable is loaded from the regular files on line 17

**Review by eli88fine — CHANGES_REQUESTED** [review]
Looks good overall. Some minor comments.  Also, Testcafe cannot be used to test time-dependent actions. It does not take the same amount of time consistently to run a test or within a test between assertions.  Please test the timeout using units tests and fakeTimers like we have done in other places.  Your Testcafe suite also is not testing that the checkmark disappears, which is another important reason to make sure to test that with unit-testing.

File: components/status/UploadFilesWidget.vue (hunk 1 above)
Comment (robinr) [reply components/status/UploadFilesWidget.vue:172]: /mantarray-frontend-components/node_modules/bootstrap/dist/css/bootstrap.css 

Has CSS property defined in the following format.
```
.bg-success {
  background-color: #28a745 !important;
}
```

In order to replace the progress-bar  background-color, as per MockflowUI, it has to be  as implemented as presently done.

```
#upload-progress-bar > .bg-success {
  background-color: #00c46f !important;
}
```

Currently this works and I don't find any issue, if this has to be very rarely used as per Mozilla documentation, then bootstrap a very popular CSS library is using !important in several part of CSS definition, why didn't the open source, popular library don't follow the inheritance rules of Mozilla documentation ?

File: tests/unit/components/status/UploadFilesWidgets.spec.js (hunk 2 above)
Comment (robinr) [reply tests/unit/components/status/UploadFilesWidgets.spec.js:57]: Corrected to `uploaded_files is 900` in the When section.

File: tests/unit/components/status/UploadFilesWidgets.spec.js (hunk 3 above)
Comment (robinr) [reply tests/unit/components/status/UploadFilesWidgets.spec.js:82]: Corrected, to have Vuex commit to happen after mount ? assume if one experiment produced 500 files and was uploaded, then the device was re-set and the old files are present in the laptop/computer's persistence storage. and an new experiment produces 400 files then would you start from 500 to 900 on the App or would you reset to start from 0 or would it start Vuex as 500 to 900 ?

File: tests/unit/store/settings.spec.js (hunk 4 above)
Comment (robinr) [reply tests/unit/store/settings.spec.js:28]: Corrected to `initialized`

**Comment by robinr** [comment]
Analyzed the E2E error, its not reproduceable so there is no fix.

File: components/status/UploadFilesWidget.vue (hunk 1 above)
Comment (eli88fine) [reply components/status/UploadFilesWidget.vue:172]: I don't know why Bootstrap is not following the Mozilla documentation.  Yesterday we agreed it was OK to use the `!important` flag in this case in our code, but please put a comment in the code describing the justification for not following the general Mozilla guidelines

Comment (robinr) [reply components/status/UploadFilesWidget.vue:172]: Included a comment with the reason for having !important.

**Review by eli88fine — APPROVED** [review]

**Pull request merged**